internal/venice-common/src/main/java/com/linkedin/venice/system/store/MetaStoreWriter.java

```diff
@@ -494,15 +500,31 @@ private void closeVeniceWriter(String metaStoreName, VeniceWriter veniceWriter,
 
   @Override
   public void close() throws IOException {
-    // Close VeniceWrites in parallel to reduce the time to shut down the server.
-    try (Timer ignore = Timer.run(
-        elapsedTimeInMs -> LOGGER.info(
-            "MetaStoreWriter takes {}ms to close {} VeniceWriters in parallel",
-            elapsedTimeInMs,
-            metaStoreWriterMap.size()))) {
-      metaStoreWriterMap.entrySet()
-          .parallelStream()
-          .forEach(entry -> closeVeniceWriter(entry.getKey(), entry.getValue(), false));
+    long startTime = System.currentTimeMillis();
+    LOGGER.info("Closing MetaStoreWriter - numOfVeniceWriters: {}", metaStoreWriterMap.size());
+    // iterate through the map and close all the VeniceWriters
+    List<CompletableFuture<VeniceResourceCloseResult>> closeFutures = new ArrayList<>(metaStoreWriterMap.size());
+
+    for (Map.Entry<String, VeniceWriter> entry: metaStoreWriterMap.entrySet()) {
+      closeFutures.add(entry.getValue().closeAsync(true));
+    }
+    CompletableFuture<Void> combinedFuture = CompletableFuture.allOf(closeFutures.toArray(new CompletableFuture[0]));
+    try {
+      combinedFuture.get(closeTimeoutMs, TimeUnit.MILLISECONDS);
+    } catch (Exception e) {
+      LOGGER.warn("Caught exception while closing VeniceWriters", e);
     }
+    // count unclosed VeniceWriters
+    int unclosedVeniceWriterCount = 0;
+    for (CompletableFuture<VeniceResourceCloseResult> future: closeFutures) {
+      if (!future.isDone()) {
+        unclosedVeniceWriterCount++;
+      }
+    }
+
+    LOGGER.info(
```

Comment:
For completed futures, it might be useful to log the stats for each `VeniceResourceCloseResult` type, if we could.

Reply:
Yeah good point